### PR TITLE
test(training): cover GRPO return and KL utils

### LINKED ISSUE #122: [Task 13] Add Unit Tests for GRPO Returns and KL Estimators
## Claim Information

- **Upstream repository**: `https://github.com/redai-infra/Relax`
- **Working fork**: `https://github.com/Dayuxiaoshui/Relax`
- **Baseline commit**: `cfda05351d4b137f60e2b71d048beab1316d9d9f`
- **Development branch**: `dev1`

## Environment

| Item | Value |
| --- | --- |
| OS | Ubuntu 24.04.4 LTS |
| Image | Current Relax development environment; the exact image tag is not exposed |
| Python | 3.12.3 |
| pytest | 9.1.1 in the project `.venv` |
| PyTorch | 2.11.0+cu130 in the project `.venv` |
| CUDA / driver | CUDA runtime 13.0 / NVIDIA driver 610.43.02; not used by these tests |
| GPU | 2 x NVIDIA GeForce RTX 5090, 32607 MiB each; test GPU usage: 0 |
| CPU | 2 x Intel Xeon Platinum 8358P, 192 physical cores / 384 logical CPUs |
| Memory | 96 GiB visible memory, no swap |

## Execution Command

This task does not use a training launch script. It adds pure CPU PyTorch unit tests and does not require a model, dataset, Ray, Megatron, CUDA, or GPU.

Required test command:

```bash
cd /data/Relax
.venv/bin/python -m pytest -v tests/utils/training/test_ppo_utils_grpo.py
```

The equivalent command when the project virtual environment is already active is:

```bash
python -m pytest -v tests/utils/training/test_ppo_utils_grpo.py
```

No new or modified runtime parameters are required.

### Related Files and Core Functions

- Source under test: `relax/utils/training/ppo_utils.py`
- New test file: `tests/utils/training/test_ppo_utils_grpo.py`
- Core function: `get_grpo_returns`
- Core function: `compute_approx_kl`

### Planned Changes

- Add at least six focused CPU-only pytest test functions.
- Call `torch.manual_seed(0)` at the beginning of every test case.
- Verify that `get_grpo_returns` broadcasts each scalar sequence reward across the shape of its corresponding per-token KL tensor.
- Cover the exact formulas for all four supported KL estimators: `k1`, `k2`, `k3`, and `low_var_kl`.
- Verify non-negativity for `k2`, `k3`, and `low_var_kl`; `k1` is intentionally excluded because its signed log-ratio estimate can be negative.
- Verify the numerical clamp applied by `low_var_kl`.
- Verify that an unknown KL estimator raises `ValueError` with a readable message.
- Use `torch.allclose(actual, expected, atol=1e-6)` for floating-point comparisons.

### Proposed Test Matrix

| Test | Expected behavior |
| --- | --- |
| GRPO reward broadcasting | Every token in each returned tensor equals that sample's scalar reward; shape and dtype follow the corresponding KL tensor |
| `k1` estimator | Returns `log_probs - log_probs_base` |
| `k2` estimator | Returns half of the squared log-ratio and is element-wise non-negative |
| `k3` estimator | Returns `exp(-log_ratio) - 1 + log_ratio` and is element-wise non-negative |
| `low_var_kl` estimator | Uses the `k3` formula and clamps the result to the documented range |
| Invalid estimator | Raises `ValueError` and includes the invalid type in the error message |

### Non-Goals

- Do not modify `get_grpo_returns`, `compute_approx_kl`, or other production code.
- Do not modify argument parsing, Controller, Service, launcher, or training scripts.
- Do not introduce new dependencies.
- Do not import or initialize Megatron, Ray, SGLang, CUDA, or distributed process groups.
- Do not add GPU, model, dataset, or end-to-end training tests.
- Do not benchmark `torch.compile` or change compilation behavior.
- Do not test unrelated PPO, GSPO, SAPO, CISPO, or REINFORCE++ utilities.

## Problem Statement

`get_grpo_returns` and `compute_approx_kl` are core pure functions used by the GRPO advantage and loss paths, but the current test suite does not contain focused unit tests for their mathematical contracts.

Without direct coverage, a change could silently break scalar reward broadcasting, alter the sign or formula of a KL estimator, remove the non-negative property of the `k2`/`k3` estimators, or accept an unsupported estimator type. Discovering these regressions only through distributed training would be slow, expensive, and difficult to diagnose.

## Proposed Solution

Add `tests/utils/training/test_ppo_utils_grpo.py` with deterministic CPU PyTorch tests that compare each function against explicitly constructed expected tensors.

The tests will use small tensors, set `torch.manual_seed(0)` in every case, and use `torch.allclose(..., atol=1e-6)` for floating-point assertions. The test module will import only PyTorch, pytest, and the two functions under test, so it can run without Megatron or CUDA.

## Alternatives Considered

1. **Cover the functions only through a GRPO training smoke test**: Rejected because it requires GPU resources and makes formula-level regressions difficult to isolate.
2. **Use exact equality for floating-point tensors**: Rejected because compiled and eager PyTorch execution can differ by harmless floating-point rounding.
3. **Assert non-negativity for all KL estimators**: Rejected because `k1` is a signed log-ratio estimator and may legitimately contain negative values.
4. **Modify the production functions while adding tests**: Out of scope because this task requests missing regression coverage, not an algorithm or API change.
5. **Add tests for every function in `ppo_utils.py`**: Rejected to keep the PR limited to the two GRPO functions named in the task.

## Use Case

When a contributor changes GRPO advantage calculation or KL handling, this focused test file provides a fast regression check before running expensive distributed training:

```python
import torch

from relax.utils.training.ppo_utils import compute_approx_kl, get_grpo_returns

rewards = torch.tensor([1.0, -0.5])
kl = [torch.zeros(3), torch.zeros(2)]
returns = get_grpo_returns(rewards, kl)

log_probs = torch.tensor([-0.2, -0.4])
log_probs_base = torch.tensor([-0.3, -0.1])
approx_kl = compute_approx_kl(log_probs, log_probs_base, "k3")
```

The resulting tensors can be checked directly against the documented formulas without loading a model or starting distributed services.

## Acceptance Criteria

- `.venv/bin/python -m pytest -v tests/utils/training/test_ppo_utils_grpo.py` passes.
- At least five non-empty test cases are included.
- Reward broadcasting and all four KL estimators are covered.
- Non-negativity is asserted for the estimators whose formulas guarantee it.
- Unknown estimator types raise `ValueError`.
- Every floating-point expectation uses `torch.allclose(..., atol=1e-6)`.
- Every test case starts with `torch.manual_seed(0)`.
- Tests run on CPU and do not depend on Megatron, Ray, SGLang, CUDA, a model, or a dataset.
- The PR contains only the new focused test file unless a test-discovered production bug is discussed separately.

## Additional Context

- Related issues or PRs: PR to be created from the `dev1` branch after Issue confirmation.
- References: Task 13 acceptance requirements for `relax/utils/training/ppo_utils.py`.
- Are you willing to contribute this feature? **Yes.**

### PR DESCRIPTION
**Pull request by Dayuxiaoshui**
## Cover GRPO return and KL utility contracts

- Validate scalar reward broadcasting to per-token return tensors
- Cover the `k1`, `k2`, `k3`, and `low_var_kl` formulas
- Check non-negative estimators, low-variance clamping, and invalid estimator errors

## What

Add `tests/utils/training/test_ppo_utils_grpo.py` with six deterministic CPU-only unit tests for:

- `get_grpo_returns`
- `compute_approx_kl`

## Why

These pure functions are used by the GRPO advantage and loss paths but did not have focused formula-level regression coverage. The new tests catch reward broadcasting, estimator formula, numerical property, and input-validation regressions without requiring distributed training.

Fixes #122

## How

- Construct small CPU tensors with explicit expected values.
- Set `torch.manual_seed(0)` at the beginning of every test.
- Compare floating-point tensors with `torch.allclose(..., atol=1e-6)`.
- Disable `torch.compile` for the formula tests so they do not require Dynamo, Inductor, Triton, or a CUDA toolchain.
- Leave the production implementation unchanged.

## Testing

```bash
CUDA_VISIBLE_DEVICES='' .venv/bin/python -m pytest -v tests/utils/training/test_ppo_utils_grpo.py
```

Result:

```text
6 passed, 14 warnings in 2.75s
```

Additional checks:

```text
git diff --cached --check: passed
python -m py_compile: passed
pre-commit run --all-files: all hooks passed except docformatter
docformatter failure: docformatter==1.3.1 imports lib2to3, which is unavailable in this Python 3.12 environment
```

- [ ] `pre-commit run --all-files` passes
- [x] Targeted tests pass (`6 passed`)
- [x] New tests added
- [ ] Documentation updated (not applicable)

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] CI/CD or build changes
- [x] Tests

## Screenshots / Logs

```text
tests/utils/training/test_ppo_utils_grpo.py::test_get_grpo_returns_broadcasts_rewards_to_token_shapes PASSED
tests/utils/training/test_ppo_utils_grpo.py::test_compute_approx_kl_k1_returns_signed_log_ratio PASSED
tests/utils/training/test_ppo_utils_grpo.py::test_compute_approx_kl_k2_matches_formula_and_is_non_negative PASSED
tests/utils/training/test_ppo_utils_grpo.py::test_compute_approx_kl_k3_matches_formula_and_is_non_negative PASSED
tests/utils/training/test_ppo_utils_grpo.py::test_compute_approx_kl_low_var_matches_formula_and_clamps_large_values PASSED
tests/utils/training/test_ppo_utils_grpo.py::test_compute_approx_kl_rejects_unknown_estimator PASSED
```

Two review rounds completed:

- Round 1 addressed Python compatibility and clarified numerical stability expectations.
- Round 2 approved the change with no P0-P3 findings.

### CODE REVIEW

**Comment by yuanlehome** [comment]
Please proceed to move the PR to open request review status.

**Comment by Dayuxiaoshui** [comment]
cc @yuanlehome

**Review by NINGBENZHE — APPROVED** [review]
lgtm
